packager/packager.cc

```diff
@@ -0,0 +1,665 @@
+                core::MutableContext ctx(gs, core::Symbols::root(), file.file);
+                auto pkg = getPackageInfo(ctx, file);
+                // getPackageInfo emits an error if the package is malformed.
+                if (pkg != nullptr) {
+                    auto &existing = packageInfoByMangledName[pkg->name.mangledName];
+                    if (existing != nullptr) {
+                        if (auto e =
+                                ctx.beginError(pkg->loc.offsets(), core::errors::Packager::RedefinitionOfPackage)) {
+                            auto pkgName = absl::StrJoin(pkg->name.fullName, "::", NameFormatter(ctx));
+                            e.setHeader("Redefinition of package `{}`", pkgName);
+                            e.addErrorLine(existing->loc, "Package `{}` originally defined here", pkgName);
+                        }
+                    } else {
+                        existing = pkg;
+                    }
+                    packageInfoByFile[file.file] = pkg;
+                    packages.emplace_back(move(pkg));
+                }
+            }
+        }
+    }
+
+    // Sort packages so that packages with the longest/most specific paths are first.
+    // That way, the first package to match a file's path is the most specific package match.
+    fast_sort(packages, [](const auto &a, const auto &b) -> bool {
+        return a->packagePathPrefix.size() > b->packagePathPrefix.size();
+    });
+
+    {
+        Timer timeit(gs.tracer(), "packager.rewritePackages");
+        // Step 2: Rewrite packages. Can be done in parallel (and w/ step 3) if this becomes a bottleneck.
+        for (auto &file : files) {
+            if (file.file.data(gs).sourceType == core::File::Type::Package) {
+                auto &packageInfo = packageInfoByFile[file.file];
+                if (packageInfo != nullptr) {
+                    core::Context ctx(gs, core::Symbols::root(), file.file);
+                    file = rewritePackage(ctx, move(file), *packageInfo, packageInfoByMangledName);
+                }
+            }
+        }
+    }
+
+    // Step 3: Find files within each package and rewrite each.
+    {
+        Timer timeit(gs.tracer(), "packager.rewritePackagedFiles");
+
+        auto resultq = make_shared<BlockingBoundedQueue<vector<ast::ParsedFile>>>(files.size());
+        auto fileq = make_shared<ConcurrentBoundedQueue<ast::ParsedFile>>(files.size());
+        for (auto &file : files) {
+            fileq->push(move(file), 1);
+        }
+
+        workers.multiplexJob("rewritePackagedFiles", [&gs, &packages, fileq, resultq]() {
+            Timer timeit(gs.tracer(), "packager.rewritePackagedFilesWorker");
+            vector<ast::ParsedFile> results;
+            u4 filesProcessed = 0;
+            ast::ParsedFile job;
+            for (auto result = fileq->try_pop(job); !result.done(); result = fileq->try_pop(job)) {
+                if (result.gotItem()) {
```

Comment:
iirc, is this the line that has been problematic in the (recent?) past, where we forget to check that `try_pop` might "succeed" but be empty, because in fact `result.done()`?

Reply:
Hmm, I don't recall that bug. This line of code will only execute if `!result.done()`.

Reply:
I have wondered if there's a better way to wrap this API so we can not think about these things.

Reply:
I think it was the cause of #3267, wasn't it?

Reply:
Oh! No. If this were #3267, the code would use `job` despite `!result.gotItem()` -- e.g., the pop failed but it's still processing `job`.